### DISCUSSION TITLE
After embedding fonts, how to make the created PDF file smaller?

### DISCUSSION THREAD

**Discussion by codingbug313** [pdfcpu/pdfcpu#971] [General]
I used the font file SourceHanSansSC-VF.ttf embedded in the PDF, and the file size is over 20 MB for a PDF with fewer than 10 pages,  I  upload font file 
Here is my code snippet：
[SourceHanSansSC-VF.zip](https://github.com/user-attachments/files/17456195/SourceHanSansSC-VF.zip)

create  new pdf  context  ：
	p.FilesInfo.CurrentFileCount = 0
			p.FilesInfo.LastFileCount = 0
			p.FilesInfo.Outfile = p.FilesInfo.FileName + "_" + strconv.Itoa(p.FilesInfo.PreIndex+1) + "." + p.FilesInfo.FileType
			pdfConfig := api.LoadConfiguration()
			pdfConfig.Optimize = true
			pdfConfig.OptimizeResourceDicts = true
			pdfConfig.OptimizeDuplicateContentStreams = true
			pdfConfig.Cmd = model.CREATE
			p.PdfProInfo.Ctx, err = pdfcpu.CreateContextWithXRefTable(pdfConfig, types.PaperSize["A4"])
			if err != nil {
				return
			}
add   new pdf  page :

func (p *CreateFileObject) GetPdfPage(mRect types.Rectangle, cropRect types.Rectangle) (pP *model.Page) {
	if len(p.PdfProInfo.Pages) <= 0 {
		page := new(model.Page)
		*page = model.NewPage(&mRect, &cropRect)
		p.PdfProInfo.Pages = append(p.PdfProInfo.Pages, nil)
		return page
	}
	if p.FilesInfo.LastFilePosY <= 0 {
		page := new(model.Page)
		*page = model.NewPage(&mRect, &cropRect)
		p.PdfProInfo.Pages = append(p.PdfProInfo.Pages, nil)
		return page
	}
	if p.FilesInfo.LastFilePosY > 0 {
		tmp := p.PdfProInfo.Pages[len(p.PdfProInfo.Pages)-1]
		if tmp == nil {
			tmp = new(model.Page)
			*tmp = model.NewPage(&mRect, &cropRect)
			//if p.FilesInfo.LastFileCount == len(p.PdfProInfo.Pages) && len(p.PdfProInfo.LastPageFm) > 0 {
			//	tmp.Fm = p.PdfProInfo.LastPageFm
			//	p.PdfProInfo.GFontMap = p.PdfProInfo.LastPageFm
			//}
		}
		return tmp
	}
	return nil
}

write pdf file :

a, b, err := create.UpdatePageTree(p.PdfProInfo.Ctx, p.PdfProInfo.Pages, p.PdfProInfo.GFontMap)
	//p.PdfProInfo.Ctx.PageCount = p.FilesInfo.CurrentFileCount
	if err != nil {
		slog.Error("UpdatePageTree ,err is :", err.Error())
		return
	}
	_, _ = a, b
	//p.Ctx.PageCount = p.FilesInfo.LastFileCount + len(p.PdfProInfo.Pages)
	//生成pdf文件
	tmpOutFile := "__" + p.FilesInfo.Outfile
	err = api.WriteContextFile(p.PdfProInfo.Ctx, filepath.Join(p.FilesInfo.FilePath, tmpOutFile))
	if err != nil {
		slog.Error("WriteContextFile ,err is :", err.Error())
		return
	}
	if err = api.ValidateFile(filepath.Join(p.FilesInfo.FilePath, tmpOutFile), nil); err != nil {
		slog.Error("ValidateFile ,err is :", err.Error())
		return
	}
	err = os.RemoveAll(filepath.Join(p.FilesInfo.FilePath, p.FilesInfo.Outfile))
	if err != nil {
		return
	}
	err = os.Rename(filepath.Join(p.FilesInfo.FilePath, tmpOutFile), filepath.Join(p.FilesInfo.FilePath, p.FilesInfo.Outfile))
	if err != nil {
		return
	}

**Comment by hhrutter** [comment]
Sorry for my late response.

CJKV fonts are really not good candidates for embedding in general because they contain a huge number of glyphs.

PDF generation using the API is really not mature.
Have you tried using JSON to create a file.
If you can manage that and it proves your point, create an issue and post the json file and I will take a look.

**Comment by hhrutter** [comment]
closed due to inactivity
